### PR TITLE
Add GEBCO data source to load_earth_relief

### PR DESCRIPTION
**Pull request by willschlitzer**
This pull request adds the GEBCO data source to `load_earth_relief`

Addresses #1786

**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If wrapping a new module, open a 'Wrap new GMT module' issue and submit reasonably-sized PRs.
- [ ] If adding new functionality, add an example to docstrings or tutorials.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

**Comment by willschlitzer** [comment]
I'm running into an error when my machine is trying to find the cached version of the GEBCO files. When I run both `grid = load_earth_relief(data_source="gebco", region = [0, 5, 10, 12])` and `grid = pygmt.grdcut(grid="@earth_gebco_01d", region = [0, 5, 10, 12])` I get the following:

```
GMTCLibError: Module 'grdcut' failed with status code 16:
grdcut [ERROR]: Libcurl Error: HTTP response code said error
grdcut [ERROR]: File /home/will/.gmt/cache/earth_gebco_01d.grd was not found
grdcut [ERROR]: Cannot find file /home/will/.gmt/cache/earth_gebco_01d.grd
grdcut [ERROR]: File /home/will/.gmt/cache/earth_gebco_01d.grd was not found
grdcut [ERROR]: Cannot find file /home/will/.gmt/cache/earth_gebco_01d.grd
grdcut [ERROR]: File /home/will/.gmt/cache/earth_gebco_01d.grd not found
```

My guess is the issue is that the Earth GEBCO files are stored under .gmt/server/earth/earth_gebco rather than .gmt/cache. Is this a GMT error for how it finds the cached files?

**Comment by seisman** [comment]
What's the output if you run the following GMT command?
```
gmt grdcut @earth_gebco_01d -R0/5/10/20 -Gout.nc -Vd
```

**Comment by willschlitzer** [comment]
> gmt grdcut @earth_gebco_01d -R0/5/10/20 -Gout.nc -Vd

Same issue. Does it work for you?

Here is the output:
```
grdcut [INFORMATION]: Processing input grid
grdcut [ERROR]: File /home/will/.gmt/cache/earth_gebco_01d.grd was not found
grdcut [ERROR]: Cannot find file /home/will/.gmt/cache/earth_gebco_01d.grd
grdcut [ERROR]: File /home/will/.gmt/cache/earth_gebco_01d.grd was not found
grdcut [ERROR]: Cannot find file /home/will/.gmt/cache/earth_gebco_01d.grd
grdcut [ERROR]: File /home/will/.gmt/cache/earth_gebco_01d.grd not found
[Session gmt (0)]: Error returned from GMT API: GMT_FILE_NOT_FOUND (16)
```

**Comment by weiji14** [comment]
I've got the same `GMT_FILE_NOT_FOUND` error when using `grdcut`, but using `gmt which -Ga @earth_gebco_01d_g` works:

```
gmt [INFORMATION]: Downloading file http://oceania.generic-mapping-tools.org/gmt_data_server.txt ...
gmt [INFORMATION]: Downloading file http://oceania.generic-mapping-tools.org/gmt_hash_server.txt ...
gmtwhich [INFORMATION]: Writing Data Table to Standard Output stream
gmtwhich [NOTICE]: Remote data courtesy of GMT data server oceania [http://oceania.generic-mapping-tools.org]

gmtwhich [NOTICE]: GEBCO Earth Relief at 1x1 arc degrees reduced by Gaussian Cartesian filtering (111.2 km fullwidth) [GEBCO Compilation Group, 2021 GEBCO 2021 Grid].
gmtwhich [NOTICE]:   -> Download grid file [105K]: earth_gebco_01d_g.grd
gmtwhich [INFORMATION]: Downloading file http://oceania.generic-mapping-tools.org/server/earth/earth_gebco/earth_gebco_01d_g.grd ...
/home/username/.gmt/server/earth/earth_gebco/earth_gebco_01d_g.grd
```

**Comment by willschlitzer** [comment]
@weiji14 Does the use of `which` request a download from the remote server regardless, or will it use the cached version of the file if available?

**Comment by weiji14** [comment]
> @weiji14 Does the use of `which` request a download from the remote server regardless or if it's cached, or will it use the cached version of the file if available? I was able to download the .grd and .nc files to .gmt/server/earth/earth_gebco.

It'll definitely take the local cached file. Looking at the verbose log though, `which` does seem to ping server network first to check for a new gmt_data_server.txt file and will pull an update if needed.

**Comment by meghanrjones** [comment]
These issues could be related to https://github.com/GenericMappingTools/gmt/pull/6160. I would be curious if it works after deleting ~/.gmt/server/gmt_data_server.txt

**Comment by willschlitzer** [comment]
> These issues could be related to [GenericMappingTools/gmt#6160](https://github.com/GenericMappingTools/gmt/pull/6160). I would be curious if it works after deleting ~/.gmt/server/gmt_data_server.txt

Just deleted both gmt_data_server.txt and gmt_data_server,tx.old; same result

**Comment by meghanrjones** [comment]
`gmt grdcut @earth_gebco_01d -R0/5/10/20 -Gout.nc -Vd` fails for me with GMT 6.3.0 but works with the main branch, due to the fix applied in https://github.com/GenericMappingTools/gmt/pull/6160. We may want to wait until after the GMT 6.4.0 release to add support for `earth_gebco` and `earth_geoid`.

**Comment by willschlitzer** [comment]
The issue I was having seems to be fixed with GMT 6.4; marking this ready for review

**Comment by willschlitzer** [comment]
How extensively should this be testing `load_earth_relief` with `data_source="gebco"`? I don't want to duplicate all of the tests for Earth relief using SRTM data, as that seems excessive, so I just added tests using 1-degree resolution with and without setting `region`.

**Comment by seisman** [comment]
> How extensively should this be testing `load_earth_relief` with `data_source="gebco"`? I don't want to duplicate all of the tests for Earth relief using SRTM data, as that seems excessive, so I just added tests using 1-degree resolution with and without setting `region`.

Yes, I agree.

File: pygmt/datasets/earth_relief.py
Comment (seisman) [inline pygmt/datasets/earth_relief.py:63]:
```suggestion
        ``"03s"`` and ``"01s"`` grids, and the missing ocean values are filled
```

File: pygmt/datasets/earth_relief.py
Comment (seisman) [inline pygmt/datasets/earth_relief.py:65]:
```suggestion
        arc-second (i.e., ``"15s"``). If True, will only load the original
```

File: pygmt/datasets/earth_relief.py
Comment (seisman) [inline pygmt/datasets/earth_relief.py:66]:
```suggestion
        land-only SRTM tiles. Only works when ``data_source="relief"``.
```

File: pygmt/datasets/earth_relief.py
Comment (seisman) [inline pygmt/datasets/earth_relief.py:73]:
```suggestion
        - **relief** : IGPP Global Earth Relief [Default]
```

File: pygmt/datasets/earth_relief.py
Comment (weiji14) [inline pygmt/datasets/earth_relief.py:64]: Should be v2.4 now according to https://web.archive.org/web/20220809113153/https://www.generic-mapping-tools.org/remote-datasets/earth-relief.html

```suggestion
        by up-sampling the SRTM15+V2.4 tiles which have a resolution of 15
```

File: pygmt/datasets/earth_relief.py
Comment (weiji14) [inline pygmt/datasets/earth_relief.py:22]: The bottom lines from L24 to L35 need to be revised as they refer to `earth_relief`, but it should also mention `earth_gebco` now, or be general enough to mention relief/gebco.

File: pygmt/datasets/earth_relief.py
Comment (weiji14) [inline pygmt/datasets/earth_relief.py:73]:
```suggestion
        - **relief** : IGPP Global Earth Relief [Default option]. See
          :gmt-datasets:`earth-relief.html`.
```

File: pygmt/datasets/earth_relief.py
Comment (weiji14) [inline pygmt/datasets/earth_relief.py:76]:
```suggestion
        - **gebco** : GEBCO Global Earth Relief with only observed relief and
          inferred relief via altimetric gravity. See
          :gmt-datasets:`earth-gebco.html`.
```

File: pygmt/datasets/earth_relief.py
Comment (weiji14) [inline pygmt/datasets/earth_relief.py:148]: Make the error a bit more informative.
```suggestion
        raise GMTInvalidInput(
            f"Invalid earth relief 'data_source' {data_source}, "
            "valid values are 'relief' and 'gebco'."
        )
```

File: pygmt/datasets/earth_relief.py
Comment (maxrjones) [reply pygmt/datasets/earth_relief.py:64]: I think it would be better to update ``"Refer to https://www.generic-mapping-tools.org/remote-datasets/earth-relief.html for more details."`` above to ``"Refer to https://www.generic-mapping-tools.org/remote-datasets/index.html for more details about available datasets, including version information and references."`` so that we don't need to update this on each SRTM15+ release and the link provided is more generic.

**Comment by seisman** [comment]
Please note that we decided to support multiple GMT versions in issue https://github.com/GenericMappingTools/pygmt/issues/1991.

As GMT 6.3.0 is still supported, and earth_gebco is known to have issues with GMT 6.3.0, I think we need to either find a fix for GMT 6.3.0 on the PyGMT side, or raise an error when earth_gebco is used with GMT 6.3.0.

**Comment by maxrjones** [comment]
> Please note that we decided to support multiple GMT versions in issue #1991.
> 
> As GMT 6.3.0 is still supported, and earth_gebco is known to have issues with GMT 6.3.0, I think we need to either find a fix for GMT 6.3.0 on the PyGMT side, or raise an error when earth_gebco is used with GMT 6.3.0.

I don't think there's a way to fix this through PyGMT. We should raise an error for earth_gebco or earth_geoid (when added) is used with GMT <6.4.

**Comment by seisman** [comment]
> > Please note that we decided to support multiple GMT versions in issue #1991.
> > As GMT 6.3.0 is still supported, and earth_gebco is known to have issues with GMT 6.3.0, I think we need to either find a fix for GMT 6.3.0 on the PyGMT side, or raise an error when earth_gebco is used with GMT 6.3.0.
> 
> I don't think there's a way to fix this through PyGMT. We should raise an error for earth_gebco or earth_geoid (when added) is used with GMT <6.4.

From comment https://github.com/GenericMappingTools/pygmt/pull/1818#issuecomment-1067996849, it seems for GMT 6.3.0, `earth_gebco_01d` doesn't work but `earth_gebco_01d_g` works, so perhaps having a default registration is a good fix.

**Comment by seisman** [comment]
xref: https://github.com/GenericMappingTools/pygmt/issues/1929

**Comment by willschlitzer** [comment]
I'm opting to just raise an error rather than try and make other data sources backwards compatible. Is there a good way to treat the GMT version like a number? The output for `Session().get_default(name="API_VERSION")` is a string, so my only fix so far has been:
```
    with Session() as lib:
        if lib.get_default(name="API_VERSION") != "6.4.0" and data_source != "igpp":
            raise GMTVersionError(
                f"The {data_source} option is not available for GMT"
                " versions before 6.4.0."
            )
```

This seems like it will cause headaches with future GMT releases. Ideally, I would like to be able to specify that a GMT version <6.4.0 would raise errors, but I'm not sure how.

File: pygmt/datasets/earth_relief.py
Comment (seisman) [inline pygmt/datasets/earth_relief.py:151]: Because `gebco` is not the default option, I prefer to rewrite these lines to:
```suggestion
    if data_source == "gebco":
        with Session() as lib:
            if Version(lib.info["version"]) < Version("6.4.0"):
                raise GMTVersionError(
                    f"The {data_source} option is not available for GMT"
                    " versions before 6.4.0."
                )
```
In this way, we can avoid the version comparison for the default cases (i.e., `data_source="igpp"`).

Comment (willschlitzer) [reply pygmt/datasets/earth_relief.py]: I went with this instead:

```
    earth_relief_sources = {"igpp": "earth_relief_", "gebco": "earth_gebco_"}
    if data_source in earth_relief_sources and data_source != "igpp":
        with Session() as lib:
            if Version(lib.info["version"]) < Version("6.4.0"):
                raise GMTVersionError(
                    f"The {data_source} option is not available for GMT"
                    " versions before 6.4.0."
                )
```

I'm not sure how other data sources will work with old versions of GMT, but I didn't want to have the only issue be raised if `gebco` is specified.

File: pygmt/datasets/earth_relief.py
Comment (seisman) [inline pygmt/datasets/earth_relief.py:78]:
```suggestion
          :gmt-datasets:`earth-relief.html`.
```

File: pygmt/tests/test_datasets_earth_relief.py
Comment (seisman) [inline pygmt/tests/test_datasets_earth_relief.py:27]:
```suggestion
def test_earth_relief_01d_igpp():
```

File: pygmt/tests/test_datasets_earth_relief.py
Comment (seisman) [inline pygmt/tests/test_datasets_earth_relief.py:29]:
```suggestion
    Test some properties of the earth relief 01d data with IGPP data.
```

File: pygmt/tests/test_datasets_earth_relief.py
Comment (seisman) [inline pygmt/tests/test_datasets_earth_relief.py:57]:
```suggestion
    Test loading low-resolution earth relief with 'region' with IGPP data.
```

File: pygmt/datasets/earth_relief.py
Comment (seisman) [inline pygmt/datasets/earth_relief.py:28]: Line 19: if `data_source="gebco", the data will be downloaded to `~/.gmt/server/earth/earth_gebco/` instead. I think we can rewrite the sentence to something like "The grids are downloaded to a user data directory (usually a subdirectory under ``~/.gmt/server/``)"

**Comment by weiji14** [comment]
Need to add `@earth_gebco_01d` to the cache too right?

**Review by seisman — APPROVED** [review]

**Comment by seisman** [comment]
@GenericMappingTools/pygmt-maintainers Can someone else give this PR a final review before we merge it?

File: pygmt/datasets/earth_relief.py
Comment (weiji14) [inline pygmt/datasets/earth_relief.py:27]: Maybe keep the note that an internet connection is needed for the first download.
```suggestion
    (usually a subdirectory under ~/.gmt/server/earth/) the first time you
    invoke this function. Afterwards, it will load the grid from the data
    directory. So you'll need an internet connection the first time around.
```

File: pygmt/datasets/earth_relief.py
Comment (weiji14) [inline pygmt/datasets/earth_relief.py:37]: Remove extra 'for more details'
```suggestion
    Refer to :gmt-datasets:`earth-relief.html` for more details about available
    datasets, including version information and references.
```

File: pygmt/datasets/earth_relief.py
Comment (weiji14) [inline pygmt/datasets/earth_relief.py:160]: Do we need to better handle the case when `use_srtm=True` and `data_source="gebco"`?

File: pygmt/datasets/earth_relief.py
Comment (weiji14) [inline pygmt/datasets/earth_relief.py:144]: Might need an `else` statement here to catch cases when users didn't use 'igpp' or 'gebco'? Perhaps move the `raise GMTInvalidInput(f"Invalid earth relief 'data_source' ...` line at the bottom here, so that the error is raised earlier.

File: pygmt/helpers/testing.py
Comment (weiji14) [inline pygmt/helpers/testing.py:164]: Suggest to move this to L156 to sort alphabetically.

File: pygmt/tests/test_datasets_earth_relief.py
Comment (weiji14) [inline pygmt/tests/test_datasets_earth_relief.py:14]: This can fit on one line
```suggestion
@pytest.mark.parametrize("data_source", ["igpp", "gebco"])
```

File: pygmt/datasets/earth_relief.py
Comment (willschlitzer) [reply pygmt/datasets/earth_relief.py]: Fixed to check for `use_srtm=True` and data_source isn't "igpp".

File: pygmt/datasets/earth_relief.py
Comment (willschlitzer) [reply pygmt/datasets/earth_relief.py]: I moved it to its own `if` statement.

File: pygmt/helpers/testing.py
Comment (willschlitzer) [reply pygmt/helpers/testing.py]: Moved.

**Review by weiji14 — APPROVED** [review]
Cool, thanks @willschlitzer!

**Pull request merged**
